google/cloud/storage/examples/storage_async_samples.cc

```diff
@@ -200,17 +202,116 @@ void ReadObjectWithOptions(
   std::cout << "The object contains " << count << " lines\n";
 }
 
+void WriteObject(google::cloud::storage_experimental::AsyncClient& client,
+                 std::vector<std::string> const& argv) {
+  //! [write-object]
+  namespace gcs = google::cloud::storage;
+  namespace gcs_ex = google::cloud::storage_experimental;
+  auto coro = [](gcs_ex::AsyncClient& client, std::string bucket_name,
+                 std::string object_name, std::string const& filename)
+      -> google::cloud::future<gcs::ObjectMetadata> {
+    std::ifstream is(filename);
+    if (is.bad()) throw std::runtime_error("Cannot read " + filename);
+    auto [writer, token] = (co_await client.WriteObject(std::move(bucket_name),
+                                                        std::move(object_name)))
+                               .value();
+    while (token.valid()) {
+      std::vector<char> buffer(1024 * 1024);
+      is.read(buffer.data(), buffer.size());
+      buffer.resize(is.gcount());
+      token = (co_await writer.Write(std::move(token),
+                                     gcs_ex::WritePayload(std::move(buffer))))
+                  .value();
+      if (is.eof()) break;
+    }
+    co_return (co_await writer.Finalize(std::move(token))).value();
+  };
+  //! [write-object]
+  // The example is easier to test and run if we call the coroutine and block
+  // until it completes.
+  auto const metadata = coro(client, argv.at(0), argv.at(1), argv.at(2)).get();
+  std::cout << "File successfully uploaded " << metadata << "\n";
+}
+
+void WriteObjectWithRetry(
+    google::cloud::storage_experimental::AsyncClient& client,
+    std::vector<std::string> const& argv) {
+  //! [write-object-with-retry]
+  namespace gcs = google::cloud::storage;
+  namespace gcs_ex = google::cloud::storage_experimental;
+  // Make one attempt to upload `is` using `writer`:
+  auto attempt =
+      [](gcs_ex::AsyncWriter& writer, gcs_ex::AsyncToken& token,
+         std::istream& is) -> google::cloud::future<google::cloud::Status> {
+    while (token.valid()) {
+      std::vector<char> buffer(1024 * 1024);
+      is.read(buffer.data(), buffer.size());
+      buffer.resize(is.gcount());
+      auto w = co_await writer.Write(std::move(token),
+                                     gcs_ex::WritePayload(std::move(buffer)));
+      if (!w) co_return std::move(w).status();
+      token = *std::move(w);
+      if (is.eof()) break;
```

Comment:
should we do `while(token.valid() && !is.eof())` ?

Reply:
It is tricky.  The value is not defined after just calling the `std::ifstream` constructor. Using a `seekg()` clears the bit and makes the code clearer.  Let's do that.